pkg/ring/token_checker.go

```diff
@@ -0,0 +1,253 @@
+	if !tc.Enabled() {
+		return tc, nil
+	}
+
+	var ctx context.Context
+	ctx, tc.cancel = context.WithCancel(context.Background())
+	go tc.watchRing(ctx)
+
+	go tc.loop()
+	return tc, nil
+}
+
+// Enabled returns whether the TokenChecker is watching the ring. Returns
+// true only when one of the CheckOn* configuration variables are
+// set to their non-zero values.
+func (tc *TokenChecker) Enabled() bool {
+	return tc.cfg.CheckOnInterval != time.Duration(0) ||
+		tc.cfg.CheckOnCreate ||
+		tc.cfg.CheckOnAppend ||
+		tc.cfg.CheckOnTransfer
+}
+
+// Shutdown stops the Token Checker. It will stop watching the ring
+// for changes and stop checking that tokens are valid on an interval.
+func (tc *TokenChecker) Shutdown() {
+	close(tc.quit)
+}
+
+// CheckToken iterates over all expected ranges and returns true
+// when the token falls within one of those ranges.
+func (tc *TokenChecker) CheckToken(token uint32) bool {
+	tc.mut.Lock()
+	defer tc.mut.Unlock()
+
+	for _, rg := range tc.expectedRanges {
+		if rg.Contains(token) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// CheckAllTokens invokes CheckToken for all current untransferred
+// tokens found in the stored IncrementalTransferer. Invokes
+// tc.InvalidTokenHandler when an invalid token was found. Returns true when
+// all tokens are valid.
+func (tc *TokenChecker) CheckAllTokens() bool {
+	toks := tc.lifecycler.incTransferer.StreamTokens()
+	invalid := []uint32{}
+
+	for _, tok := range toks {
+		valid := tc.CheckToken(tok)
+		if !valid {
+			invalid = append(invalid, tok)
+		}
+	}
+
+	numInvalid := len(invalid)
+
+	if tc.UnexpectedTokenHandler != nil {
+		tc.UnexpectedTokenHandler(invalid)
+	}
+
+	return numInvalid == 0
+}
+
+// watchRing watches the ring and updates the set of expected tokens
+// for a TokenChecker.
+func (tc *TokenChecker) watchRing(ctx context.Context) {
+	numIngesters := 0
+
+	tc.lifecycler.KVStore.WatchKey(ctx, ConsulKey, func(v interface{}) bool {
+		if v == nil {
+			level.Info(util.Logger).Log("msg", "ring doesn't exist in consul yet")
+			return true
+		}
+
+		level.Debug(util.Logger).Log("msg", "updating token checker ring definition")
+
+		desc := v.(*Desc)
+		desc.Tokens = migrateRing(desc)
+		tc.mut.Lock()
+		tc.ring = desc
+		tc.mut.Unlock()
+		tc.updateExpectedRanges()
+
+		if len(desc.Ingesters) != numIngesters {
+			level.Debug(util.Logger).Log(
+				"msg", "number of ingesters in ring grew",
+				"ring_definition", desc.String(),
+			)
+			numIngesters = len(desc.Ingesters)
+		}
+
+		return true
+	})
+}
+
+// updateExpectedRanges goes through the ring and finds all expected ranges
+// given the current set of tokens in a Lifecycler.
+func (tc *TokenChecker) updateExpectedRanges() {
+	tokens := tc.lifecycler.getTokens()
+	expected := []TokenRange{}
+
+	for _, tok := range tokens {
+		for replica := 0; replica < tc.lifecycler.cfg.RingConfig.ReplicationFactor; replica++ {
+			endRanges, err := tc.ring.Predecessors(NeighborOptions{
```

Comment:
Yes, I agree that it's really expensive. I've tried many attempts to implement something less computationally complex for `Predecessors`, but couldn't in a reasonable amount of time. 

The primary issue is that when we search forwards through the ring, we only accept the _first_ token from each unique ingester. If we search backwards, we're only accepting the _last_ token from each unique ingester. I wasn't able to come up with an algorithm that searches backwards and also accounts for this difference. 

If you have any ideas on how else we can find all of the replica sets a token is in, please let me know, I'd like to make this better.

Reply:
It may help if we only run `updateExpectedRanges` when the set of tokens change. Right now, it's going to run with every heartbeat.

Reply:
I've come up with a way to improve `Predecessors` to be around `O(offset)`, in the average case depending on how the ring is arranged, and will make this operation a little less expensive, roughly `O(#ingesters * #tokens_per_ingester * replicationFactor^2)`. I'll implement it tomorrow and tag you so you can take a look.

Reply:
I was just thinking about the same issue... perhaps some kind of ring cleanup at the beginning would help? It would remove tokens for unhealthy ingesters, merge adjacent tokens for same ingesters [not 100% sure this is safe, but it seems like it should be]. That would simplify the data, and could possibly make code navigating it simpler as well?